### PR TITLE
[Clang-tidy] Run clang-tidy on modified code

### PR DESCRIPTION
**Pull request by troopa81**
This PR proposes to run clang-tidy on modified/added code for every PR. 

I removed the CMake option WITH_CLANG_TIDY because it takes way too much time to then compile QGIS (reached Github 6 hours timeout when trying). In order to reproduce locally, I added a script clang-tidy.sh that runs clang-tidy on some files with the appropriate options (it just needs to have the CMAKE_EXPORT_COMPILE_COMMANDS defined.

I reduced the number of check categories to *bugprone-*, *clang-diagnostic-*, *clang-analyzer-* (the two last one are default in clang-tidy, in order to not raise too much non-relevant checks. But the content of clangtidy_checks.txt is opened to discussions.

### CODE REVIEW

**Comment by elpaso** [comment]
Nice! Should we add it to the pre-commit script too?

**Comment by troopa81** [comment]
> Nice! Should we add it to the pre-commit script too?

IMHO no, because It's very slow. As I understand it, it compiles the source and then use the produced [AST](https://clang.llvm.org/docs/IntroductionToTheClangAST.html) to do a thorough analysis according to given checks. 

And the more checks you add, the slower it is!

**Comment by troopa81** [comment]
Unrelated test failure in mingw

Project  settings would need to be updated in order to remove the *2 expected checks* because I added a *with-compile-commands* option.

File: .docker/docker-qgis-clangtidy.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+set -e
+
+SRCDIR=${CTEST_SOURCE_DIR-/root/QGIS}
+cd ${SRCDIR}
+
+# This is needed for `git status` to work, see
+# https://github.com/qgis/QGIS/runs/6733585841?check_suite_focus=true#step:13:89
+git config --global --add safe.directory ${SRCDIR}
+
+echo "::group::Install cland tidy"
```
Comment (nyalldawson) [inline .docker/docker-qgis-clangtidy.sh:26]:
```suggestion
echo "::group::Install clang tidy"
```

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Let's give it a shot, and see how well it performs in practice :+1:

**Pull request merged**